src/main/java/org/apache/sysds/runtime/instructions/fed/CtableFEDInstruction.java

```diff
@@ -0,0 +1,331 @@
+
+			fedMap.getFederatedRanges()[i].setBeginDim(1, i == 0 ? 0 : fedMap.getFederatedRanges()[i-1].getBeginDims()[1]);
+			fedMap.getFederatedRanges()[i].setEndDim(1, dims2[i]);
+		}
+
+		// set output
+		out.getDataCharacteristics().set(d1, d2, (int) mo1.getBlocksize(), mo1.getNnz());
+		out.setFedMapping(fedMap.copyWithNewID(fr.getID()));
+
+		long varID = FederationUtils.getNextFedDataID();
+		out.getFedMapping().mapParallel(varID, (range, data) -> {
+			try {
+				FederatedResponse response = data.executeFederatedOperation(new FederatedRequest(
+					FederatedRequest.RequestType.EXEC_UDF, -1,
+					new SliceOutput(data.getVarID(), fedSize))).get();
+				if(!response.isSuccessful())
+					response.throwExceptionFromResponse();
+			}
+			catch(Exception e) {
+				throw new DMLRuntimeException(e);
+			}
+			return null;
+		});
+	}
+
+	private MatrixBlock aggResult(Future<FederatedResponse>[] ffr) {
+		MatrixBlock resultBlock = new MatrixBlock(1, 1, 0);
+		int dim1 = 0, dim2 = 0;
+		for(int i = 0; i < ffr.length; i++) {
+			try {
+				MatrixBlock mb = ((MatrixBlock) ffr[i].get().getData()[0]);
+				dim1 = mb.getNumRows()  > dim1 ? mb.getNumRows() : dim1;
+				dim2 = mb.getNumColumns()  > dim2 ? mb.getNumColumns() : dim2;
+
+				// set next and prev to same output dimensions
+				MatrixBlock prev = new MatrixBlock(dim1, dim2, 0.0);
+				prev.copy(0, resultBlock.getNumRows()-1, 0, resultBlock.getNumColumns()-1, resultBlock, true);
+
+				MatrixBlock next = new MatrixBlock(dim1, dim2, 0.0);
+				next.copy(0, mb.getNumRows()-1, 0, mb.getNumColumns()-1, mb, true);
+
+				// add worker results
+				BinaryOperator plus = InstructionUtils.parseBinaryOperator("+");
+				resultBlock = prev.binaryOperationsInPlace(plus, next);
+			}
+			catch(Exception e) {
+				e.printStackTrace();
+			}
+		}
+		return resultBlock;
+	}
+
+	private Long[] getOutputDimension(ExecutionContext ec, CPOperand inOp, CPOperand outOp, FederatedRange[] federatedRanges) {
+		MatrixObject in = ec.getMatrixObject(inOp);
+		Long[] fedDims = new Long[federatedRanges.length];
+
+		if(!in.isFederated()) {
+			//slice
+			MatrixBlock mb = in.acquireReadAndRelease();
+			IntStream.range(0, federatedRanges.length).forEach(i -> {
+				MatrixBlock sliced = mb
+					.slice(federatedRanges[i].getBeginDimsInt()[0], federatedRanges[i].getEndDimsInt()[0] - 1);
+				fedDims[i] = (long) sliced.max();
+			});
+			return fedDims;
+		}
+
+		// construct new instString
+		String maxInstrString = instString.replace("ctable", "uamax");
+		String[] instParts = maxInstrString.split(Lop.OPERAND_DELIMITOR);
+		String[] maxInstParts = new String[] {instParts[0], instParts[1],
+			InstructionUtils.concatOperandParts(inOp.getName(), DataType.MATRIX.name(), (ValueType.FP64).name()),
+			InstructionUtils.concatOperandParts(outOp.getName(), DataType.SCALAR.name(), (ValueType.FP64).name()), "16"};
+		maxInstrString = String.join(Lop.OPERAND_DELIMITOR, maxInstParts);
```

Comment:
It improves readability if something like this is put into a separate method with the name you have written in the comment. It would make this method shorter and it would be easier to see which steps the method consists of.